src/main/cpp/src/datetime_parser.cu

```diff
@@ -0,0 +1,527 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <map>
+#include <numeric>
+#include <vector>
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/copying.hpp>
+#include <cudf/detail/null_mask.hpp>
+
+#include <cudf/scalar/scalar.hpp>
+#include <cudf/search.hpp>
+#include <cudf/strings/string_view.cuh>
+#include <cudf/strings/strings_column_view.hpp>
+#include <cudf/types.hpp>
+#include <cudf/utilities/default_stream.hpp>
+#include <cudf/wrappers/timestamps.hpp>
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/optional.h>
+#include <thrust/transform.h>
+#include <thrust/tuple.h>
+
+#include "datetime_parser.hpp"
+
+namespace
+{
+
+/**
+ * represents local date time in a time zone.
+ */
+struct timestamp_components
+{
+  int32_t year; // max 6 digits
+  int8_t month;
+  int8_t day;
+  int8_t hour;
+  int8_t minute;
+  int8_t second;
+  int32_t microseconds;
+};
+
+/**
+ * convert a local time in a time zone to UTC timestamp
+ */
+__device__ __host__ thrust::tuple<cudf::timestamp_us, bool>
+to_utc_timestamp(timestamp_components components, cudf::string_view const &time_zone)
+{
+  // TODO replace the fake implementation
+  long seconds = components.year * 365L * 86400L +
+                 components.month * 30L * 86400L +
+                 components.day * 86400L +
+                 components.hour * 3600L +
+                 components.minute * 60L +
+                 components.second;
+  long us = seconds * 1000000L + components.microseconds;
+  return thrust::make_tuple(cudf::timestamp_us{cudf::duration_us{us}}, true);
+}
+
+__device__ __host__ inline bool is_whitespace(const char chr)
+{
+  switch (chr)
+  {
+  case ' ':
+  case '\r':
+  case '\t':
+  case '\n':
+    return true;
+  default:
+    return false;
+  }
+}
+
+// compare 2 strings are equal ignore case, the expect string should be lower-case
+__device__ __host__ inline bool equals(const char *actual_begin, const char *actual_end,
+                                        const char *expect_begin, const char *expect_end)
+{
+  if (actual_end - actual_begin != expect_end - expect_begin)
+  {
+    return false;
+  }
+
+  while (actual_begin < actual_end)
+  {
+    // the diff between upper case and lower case for a same char is 32
+    if (*actual_begin != *expect_begin && *actual_begin != (*expect_begin - 32))
```

Comment:
This only works for ASCII characters and only for a-z. Ideally we would rename this to equalsASCIIIgnoreCase at least. Ideally it would restrict the case check to only A-Z.

Reply:
Done.